src/Storages/StorageFile.cpp

```diff
@@ -596,19 +572,9 @@ namespace
                     auto file_enumerator = archive_reader->firstFile();
```

Comment:
So, as I understood, to get back this file enumerator on next iteration, we will need the unique ptr to read buffer we returned on previous iteration (like it's done in StorageFileSource). I think we can do similar here. For example add new method to `IReadBufferIterator` like `setPreviousReadBuffer` or `returnReadBuffer` (I am not good at naming) that will be called before `next()` in `readSchemaFromFormat`. Or maybe add argument `unique_ptr<ReadBuffer> prev_buf = nullptr` to `next()` method, so we will get previously returned buffer as an argument.

Reply:
I'll use `setPreviousReadBuffer` because default values and virtual functions don't work well together. 
Not sure I was afraid to change `IReadBufferIterator`, probably not to mess up something.